target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_avm_fritz3390.dts

```diff
@@ -0,0 +1,338 @@
+/dts-v1/;
+
+#include "vr9.dtsi"
+
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/mips/lantiq_rcu_gphy.h>
+
+/ {
+	compatible = "avm,fritz3390", "lantiq,xway", "lantiq,vr9";
+	model = "AVM FRITZ!Box 3390";
+
+	chosen {
+		bootargs = "console=ttyLTQ0,115200";
+	};
+
+	aliases {
+		led-boot = &led_power_green;
+		led-failsafe = &led_power_red;
+		led-running = &led_power_green;
+		led-upgrade = &led_power_red;
+
+		led-dsl = &led_dsl;
+		led-internet = &led_info;
+		led-wifi = &led_wifi;
+	};
+
+	memory@0 {
+		device_type = "memory";
+		reg = <0x0 0x8000000>;
+	};
+
+	keys {
+		compatible = "gpio-keys-polled";
+		poll-interval = <100>;
+
+		power {
+			label = "power";
+			gpios = <&gpio 1 GPIO_ACTIVE_HIGH>;
+			linux,code = <KEY_POWER>;
+		};
+
+		wifi {
+			label = "wlan";
+			gpios = <&gpio 29 GPIO_ACTIVE_HIGH>;
+			linux,code = <KEY_WLAN>;
+		};
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_power_green: power_green {
+			label = "fritz3390:green:power";
+			gpios = <&gpio 45 GPIO_ACTIVE_LOW>;
+			default-state = "keep";
+		};
+
+		led_power_red: power_red {
+			label = "fritz3390:red:power";
+			gpios = <&gpio 46 GPIO_ACTIVE_LOW>;
+		};
+
+		led_wifi: wifi {
+			label = "fritz3390:green:wlan";
+			gpios = <&gpio 36 GPIO_ACTIVE_LOW>;
+		};
+
+		led_dsl: dsl {
+			label = "fritz3390:green:dsl";
+			gpios = <&gpio 35 GPIO_ACTIVE_LOW>;
+		};
+
+		led_lan {
+			label = "fritz3390:green:lan";
+			gpios = <&gpio 47 GPIO_ACTIVE_LOW>;
+		};
+
+		led_info: info {
+			label = "fritz3390:green:info";
+			gpios = <&gpio 33 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	gpio-export {
+		compatible = "gpio-export";
+
+		gpio_wasp_reset {
+			gpio-export,name = "fritz3390:wasp:reset";
```

Comment:
I assume that this is for the "other SoC"?

my personal on this topic:
I would work with the upstream linux maintainers to see how this can be integrated upstream (instead of going the route by getting it into OpenWrt without upstream backing). I have updated the lantiq target to newer kernel versions in the past and it was a pain. This is why I am upstreaming lantiq code whenver I have time (my latest mainline contribution for lantiq is the PCIe PHY driver in v5.4: ```drivers/phy/lantiq/phy-lantiq-vrx200-pcie.c```)

I don't know how communication with this "other SoC" works (I just read that it's some Atheros / QCA part).
However, there is a subsystem for "remote processors" in ```drivers/remoteproc/```
My high-level idea (again, without understanding the details - I think examples are better than long text) would look like this in .dts:
`
  &mdio {
    wasp: remote-processor@7 {
      comatible = "qca,wasp-remote-processor";
      reg = <0x7>;
      reset-gpios = <&gpio 34 GPIO_ACTIVE_HIGH>;
      firmware-name = "fritzbox3390-wasp-remote-processor-firmware"; /* optional */
    }
  };
`
(I don't know how sharing the configuration between the lantiq host processor and the QCA remote processor would work in that case. that's one of the reasons why I think it's important that this is discussed upstream. at least there are cases where ralink SoCs are connected similarly - so you could be the first to implement something that other drivers will re-use :))

Reply:
Yes, it's for the Atheros SoC. The communication works as follows:

  * Atheros is brought out of Reset by driving the reset GPIO high
  * a stage 1 firmware is uploaded via MDIO (called "ath_tgt_fw1.fw" in the original firmware); I'm uploading the original firmware for now
  * the stage 1 loader broadcasts ethernet frames and waits for the stage 2 upload which contains an actual Linux kernel.
  * the Atheros WiFi interface is exposed via the internal Gigabit connection to the Lantiq and looks like a native interface on the Lantiq. Basically, it looks a lot like an ath10k-htc implementation, but with a Gigabit connection instead of a USB connection. I went a slightly different approach here and uploaded a separate, complete OpenWrt initramfs image built for an ath79 target.

I implemented as much as possible in userspace, since the protocol for the upload is still a work in progress (I also started working on a 4490, it's very similar but requires some adaptations). In case you want to get a more thorough insight, the uploader code can be found at https://github.com/andyboeh/wasp_uploader.

Reply:
So I had a quick look into the remoteproc interface: To me, it seems as if it was meant for communicating with very specific cores or firmwares. While it would be suitable for supporting the WiFi functionality only, it would require a specific firmware to be written for the ath79 target (similar to the ath9k_htc firmware) - this is way over my head.

On the other hand, my approach leads to a fully customizable second OpenWrt instance on the same board which IMHO greatly improves the capability of the device.

Reply:
my understanding is that the remoteproc subsystem is used for booting "remote CPUs"
there's no definition how to communicate with the software (can be OpenWrt, some RTOS, DSP firmware, etc.).

but as you discovered: my idea is not perfect
that's why I suggested to discuss this with the maintainers of the Linux kernel (I think the remoteproc, general networking, wireless and MIPS subsystems are relevant in this discussion)

Reply:
I overslept it a few nights: Currently, I don't feel very comfortable with the firmware upload code. It's entirely reverse engineered, since I couldn't find any documentation on it. Thus, I would prefer a user-space solution that allows easier updates than an in-kernel solution. Plus, the 3490 uses a similar concept, but uses different registers which makes things more complex.